### PR TITLE
Reset the native renderer only when the GL thread exits

### PR DESCRIPTION
**Pull request by LukasPaczos**
Closes https://github.com/mapbox/mapbox-gl-native/issues/14618.
Closes https://github.com/mapbox/mapbox-gl-native/issues/14592.
Closes https://github.com/mapbox/mapbox-gl-native/issues/14252.
Closes https://github.com/mapbox/mapbox-gl-native/issues/14463.
Closes https://github.com/mapbox/mapbox-gl-native/issues/14554. It still blocks when scrolling a map inside a recycler, which is unavoidable because we are detaching the view, but it doesn't block for `#onStop` anymore.
Also doesn't force tiles to reload after coming back from the background (refs https://github.com/mapbox/mapbox-gl-native/pull/12844#issuecomment-493759011).

Hopefully, this should also replace https://github.com/mapbox/mapbox-gl-native/pull/14821. Since we are not resetting the native renderer when the `GLSurfaceView` is still able to render, we are not going to run into a missing pointer scenario. When running the introduced in https://github.com/mapbox/mapbox-gl-native/pull/14821 test, I'm seeing:
```
Abort message: '../../../../../../../src/mbgl/gl/renderer_backend.cpp:31: void mbgl::gl::RendererBackend::assumeFramebufferBinding(const gl::FramebufferID): assertion "gl::value::BindFramebuffer::Get() == getContext<gl::Context>().bindFramebuffer.getCurrentValue()" failed'
```
Can this just be an issue with the test itself, where we're trying to force the surface to draw when it wouldn't in real-world case? Otherwise, I can reintroduce the `hasSurface` flag just to cover this one edge case, still without resetting the pointer.

All of the previously introduced integration tests are passing for me, but if we'd like to move forward with this, a device farm test run would be great.

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java [hunk 1]
```diff
@@ -801,10 +802,10 @@ public void removeOnStyleImageMissingListener(@NonNull OnStyleImageMissingListen
 
   /**
    * Set a callback that's invoked when map needs to release unused image resources.
-   *
+   * <p>
    * A callback will be called only for unused images that were provided by the client via
    * {@link OnStyleImageMissingListener#onStyleImageMissing(String)} listener interface.
-   *
+   * <p>
```
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: `</p>`

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java [hunk 2]
```diff
@@ -816,7 +817,7 @@ public void addOnCanRemoveUnusedStyleImageListener(@NonNull OnCanRemoveUnusedSty
 
   /**
    * Removes a callback that's invoked when map needs to release unused image resources.
-   *
+   * <p>
```
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: needs a closing `</p>`

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/glsurfaceview/MapboxGLSurfaceView.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    if (detachedListener != null) {
+      detachedListener.onGLSurfaceViewDetached();
+    }
+    super.onDetachedFromWindow();
+  }
+
+  @Override
+  protected void onAttachedToWindow() {
+    super.onAttachedToWindow();
+  }
```
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/glsurfaceview/MapboxGLSurfaceView.java]: no-op override can be removed

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/glsurfaceview/MapboxGLSurfaceView.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Set a listener that gets notified when the view is detached form window.
+   *
+   * @param detachedListener listener
+   */
+  public void setDetachedListener(OnGLSurfaceViewDetachedListener detachedListener) {
```
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/renderer/glsurfaceview/MapboxGLSurfaceView.java]: `@NonNull`

**Comment by yandroidUA** [comment]
Hi all, same issue:
A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0 in tid 3266
